sql-plugin/src/main/scala/com/nvidia/spark/rapids/FoldLocalAggregate.scala

```diff
@@ -0,0 +1,137 @@
+      case LocalAggregatePattern(finalAgg: BaseAggregateExec, partAgg: BaseAggregateExec) =>
+        // Spark eliminates the filter for the aggExpressions in Final mode. So, we need to copy
+        // the filter from the corresponding partial aggExpressions.
+        val aggExpressions = finalAgg.aggregateExpressions.zip(
+          partAgg.aggregateExpressions).map { case (finalAggExpr, partAggExpr) =>
+            require(finalAggExpr.filter.isEmpty, "FinalAggExpression should have no filter")
+            finalAggExpr.copy(mode = Complete, filter = partAggExpr.filter)
+        }
+        val aggAttributes = aggExpressions.map(_.resultAttribute)
+        finalAgg match {
+          case hash: HashAggregateExec =>
+            hash.copy(
+              // Need to use partial group expressions. Otherwise, groupBy(Literal) will fail.
+              groupingExpressions = partAgg.groupingExpressions,
+              aggregateExpressions = aggExpressions,
+              aggregateAttributes = aggAttributes,
+              child = partAgg.child)
+          case sort: SortAggregateExec =>
+            sort.copy(
+              // Need to use partial group expressions. Otherwise, groupBy(Literal) will fail.
+              groupingExpressions = partAgg.groupingExpressions,
+              aggregateExpressions = aggExpressions,
+              aggregateAttributes = aggAttributes,
+              child = partAgg.child)
+          case obj: ObjectHashAggregateExec =>
+            obj.copy(
+              // Need to use partial group expressions. Otherwise, groupBy(Literal) will fail.
+              groupingExpressions = partAgg.groupingExpressions,
+              aggregateExpressions = aggExpressions,
+              aggregateAttributes = aggAttributes,
+              child = partAgg.child)
+          case _ =>
+            throw new IllegalStateException(s"Unexpected AggregateExec: $finalAgg")
+        }
+    }
+  }
+}
+
+/**
+ * Capture the pattern of LocalAggregate, which consists of two-stage AggregateExecs being
+ * connected with each other directly (without the ShuffleExchange in between):
+ * +- FinalAggregateExec
+ *    +- PartialAggregateExec
+ *
+ * The LocalAggregate can be emerged regardless HashAggregateExec, SortAggregateExec or
+ * ObjectHashAggregateExec.
+ */
+object LocalAggregatePattern extends Logging {
+  def unapply(plan: SparkPlan): Option[(BaseAggregateExec, BaseAggregateExec)] = {
+    plan match {
+      case hashAgg: HashAggregateExec
+        if hashAgg.child.isInstanceOf[HashAggregateExec] &&
+          isLocalExchange(hashAgg, hashAgg.child.asInstanceOf[HashAggregateExec]) =>
+        Some(hashAgg -> hashAgg.child.asInstanceOf[HashAggregateExec])
+      case sort: SortAggregateExec
+        if sort.child.isInstanceOf[SortAggregateExec] &&
+          isLocalExchange(sort, sort.child.asInstanceOf[SortAggregateExec]) =>
+        Some(sort -> sort.child.asInstanceOf[SortAggregateExec])
+      case obj: ObjectHashAggregateExec
```

Comment:
Do we have test for this `ObjectHashAggregateExec` ?

Reply:
Added one.